finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/JTreeMapBackedHeaderMap.scala

```diff
@@ -0,0 +1,79 @@
+package com.twitter.finagle.http.headers
+
+import com.twitter.finagle.http.HeaderMap
+import scala.annotation.tailrec
+
+/**
+ * Mutable, thread-safe [[HeaderMap]] implementation, backed by
+ * a mutable [[Map[String, Header]]], where the map key
+ * is forced to lower case
+ */
+final private[http] class JTreeMapBackedHeaderMap extends JMapBackedHeaderMap {
+  
+  override val underlying: java.util.TreeMap[String, Header.Root] = 
+    new java.util.TreeMap[String, Header.Root](JTreeMapBackedHeaderMap.SharedComparitor)
+
+    def getAll(key: String): Seq[String] = underlying.synchronized {
+      underlying.get(key.toLowerCase) match {
+        case null => Nil
+        case r: Header.Root => r.values
+      }
+    }
+
+  // Does not validate key and value.
+  def addUnsafe(key: String, value: String): this.type  = underlying.synchronized {
+    def header = Header.root(key, value)
+    underlying.get(key) match {
+      case null => underlying.put(key, header)
+      case h    => h.add(key, value)
+    }
+    this
+  }
+
+  // Does not validate key and value.
+  def setUnsafe(key: String, value: String): this.type = underlying.synchronized {
+    underlying.put(key, Header.root(key, value))
+    this
+  }
+
+  def get(key: String): Option[String] = underlying.synchronized {
+    Option(underlying.get(key)).map(_.value)
```

Comment:
Let's do a `match` to avoid allocating a second `Some` and the lambda used to `map`.